### PR TITLE
[BEAM-1805] a new option `asInnerJoin` for CoGroupByKey

### PR DESCRIPTION
**Pull request by XuMingmin**
*INNER_JOIN* is available with [Join](https://github.com/apache/beam/blob/master/sdks/java/extensions/join-library/src/main/java/org/apache/beam/sdk/extensions/joinlibrary/Join.java). However it's limited to two `PCollection`s, it's helpful to add it to `CoGroupByKey` for the case with multiple `PCollection` joined.

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @davorbonaci

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java [hunk 1]
```diff
@@ -74,6 +75,16 @@
 public class CoGroupByKey<K> extends
     PTransform<KeyedPCollectionTuple<K>,
                PCollection<KV<K, CoGbkResult>>> {
+
+  /**
+   *  options to filter output results.
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: I'd like to improve this Javadoc to include full description and specification here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java [hunk 2]
```diff
@@ -74,6 +75,16 @@
...
+   *
+   */
+  public enum FilterOption {
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: Can this be `private` instead?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java [hunk 3]
```diff
@@ -84,8 +95,19 @@
     return new CoGroupByKey<>();
   }
 
+  /**
+   * Add a {@link FilterOption#INNER_JOIN} to output {@code PCollection} of {@code CoGroupByKey}.
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: I'd like to improve this Javadoc to include full description and specification here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java [hunk 4]
```diff
@@ -166,6 +193,35 @@ private CoGroupByKey() { }
   }
 
   /**
+   * A filter act as INNER_JOIN in SQL.
+   *
+   * @param <K>
+   */
+  private static class InnerJoinFilter<K>
+      extends PTransform<PCollection<KV<K, CoGbkResult>>, PCollection<KV<K, CoGbkResult>>> {
+    private static final long serialVersionUID = 5338015894726734137L;
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: This shouldn't be necessary; please cut.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java [hunk 5]
```diff
@@ -166,6 +193,35 @@ private CoGroupByKey() { }
...
+
+    @Override
+    public PCollection<KV<K, CoGbkResult>> expand(PCollection<KV<K, CoGbkResult>> input) {
+      return input.apply("InnerJoinFilter",
+          ParDo.of(new DoFn<KV<K, CoGbkResult>, KV<K, CoGbkResult>>() {
+            private static final long serialVersionUID = -2025094423705642840L;
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: This shouldn't be necessary; please cut.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java (hunk 4 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: +1

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java (hunk 5 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: +1

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java (hunk 2 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/CoGroupByKey.java]: only need to expose for test, may be defined as 
`@VisibleForTesting static enum FilterOption`

**Comment by davorbonaci** [comment]
@robertwb, any thoughts perhaps?

**Comment by robertwb** [comment]
Would it be preferable instead to augment the join library to handle more than two PCollections. I'm wary of a huge number of features being added to CoGroupByKey (especially as it it set to become the grouping primitive in lieu of "ordinary" GroupByKey). 

An alternative would be to expose this FilterFn that a user could apply immediately following a CoGBK. (One could also imagine options that require at most (or even exactly) one element per iterable, and unwrap them.)

**Comment by XuMingmin** [comment]
Good point @robertwb 

Agree to separate the filter options from core code of `CoGroupByKey`, let me move it to a separate class like `CoGbkResultFilterFn`, can add outer_join later.

The ideal way is to update `join-library`. It's more complex as `join-library` uses KV in the implementation which is fixed with two sources, would see how to handle it in another task.

**Comment by XuMingmin** [comment]
create BEAM-1987 to support multiple sources in join-library as well.

**Comment by XuMingmin** [comment]
close this PR, and move to fix in BEAM 1987

**Pull request closed without merging**